include/GafferScene/SceneAlgo.inl

```diff
@@ -0,0 +1,156 @@
+//  LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING
+//  NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS
+//  SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+//
+//////////////////////////////////////////////////////////////////////////
+
+#include "tbb/task.h"
+
+namespace GafferScene
+{
+
+namespace Detail
+{
+
+template <class ThreadableFunctor>
+class TraverseTask : public tbb::task
+{
+
+	public :
+
+		TraverseTask(
+			const GafferScene::ScenePlug *scene,
+			const Gaffer::Context *context,
+			ThreadableFunctor &f
+		)
+			:	m_scene( scene ), m_context( context ), m_f( f )
+		{
+		}
+
+		virtual ~TraverseTask()
+		{
+		}
+
+		virtual task *execute()
+		{
+
+			Gaffer::ContextPtr context = new Gaffer::Context( *m_context, Gaffer::Context::Borrowed );
+			context->set( ScenePlug::scenePathContextName, m_path );
+			Gaffer::Context::Scope scopedContext( context.get() );
+
+			if( m_f( m_scene, m_path ) )
+			{
+				IECore::ConstInternedStringVectorDataPtr childNamesData = m_scene->childNamesPlug()->getValue();
+				const std::vector<IECore::InternedString> &childNames = childNamesData->readable();
+
+				set_ref_count( 1 + childNames.size() );
+
+				ScenePlug::ScenePath childPath = m_path;
+				childPath.push_back( IECore::InternedString() ); // space for the child name
+				for( std::vector<IECore::InternedString>::const_iterator it = childNames.begin(), eIt = childNames.end(); it != eIt; it++ )
+				{
+					childPath[m_path.size()] = *it;
+					TraverseTask *t = new( allocate_child() ) TraverseTask( *this, childPath );
+					spawn( *t );
+				}
+				wait_for_all();
+			}
+
+			return NULL;
+		}
+
+	protected :
+
+		TraverseTask( const TraverseTask &other, const ScenePlug::ScenePath &path )
+			:	m_scene( other.m_scene ),
+				m_context( other.m_context ),
+				m_f( other.m_f ),
+				m_path( path )
+		{
+		}
+
+	private :
+
+		const GafferScene::ScenePlug *m_scene;
+		const Gaffer::Context *m_context;
+		ThreadableFunctor &m_f;
+		GafferScene::ScenePlug::ScenePath m_path;
+
+};
+
+template <class ThreadableFunctor>
+struct ThreadableFilteredFunctor
+{
+    ThreadableFilteredFunctor( ThreadableFunctor &f, const Gaffer::IntPlug *filter ): m_f( f ), m_filter( filter ){}
+
+    bool operator()( const GafferScene::ScenePlug *scene, const GafferScene::ScenePlug::ScenePath &path )
+    {
+		const GafferScene::Filter::Result match = (GafferScene::Filter::Result)m_filter->getValue();
+		if( match & GafferScene::Filter::ExactMatch )
+		{
+			m_f( scene, path );
+		}
+
+		return ( match & GafferScene::Filter::DescendantMatch ) != 0;
+    }
+
+	ThreadableFunctor &m_f;
+	const Gaffer::IntPlug *m_filter;
+
+};
+
+} // namespace
+
+template <class ThreadableFunctor>
+void parallelTraverse( const GafferScene::ScenePlug *scene, ThreadableFunctor &f, const Gaffer::Context *context )
+{
+	Gaffer::ContextPtr c = new Gaffer::Context( *Gaffer::Context::current(), Gaffer::Context::Borrowed );
```

Comment:
This seems to be ignoring the context argument (the one I want to remove) - highlighting quite nicely the ambiguity that occurs when an external API call takes a context explicitly.